shark-hprof/src/main/java/shark/HprofDeobfuscator.kt

```diff
@@ -0,0 +1,237 @@
+          hprofVersion = hprof.hprofVersion
+        ).use { writer ->
+          reader.readHprofRecords(setOf(HprofRecord::class),
+            OnHprofRecordListener { _,
+              record ->
+              // HprofWriter automatically emits HeapDumpEndRecord, because it flushes
+              // all continuous heap dump sub records as one heap dump record.
+              if (record is HeapDumpEndRecord) {
+                return@OnHprofRecordListener
+              }
+
+              when (record) {
+                is StringRecord -> {
+                  writer.write(
+                    createDeobfuscatedStringRecord(record, proguardMapping, hprofStringCache)
+                  )
+                }
+                is HeapDumpRecord -> {
+                  when (record) {
+                    is ObjectRecord -> {
+                      when (record) {
+                        is ClassDumpRecord -> {
+                          val recordsToWrite = createDeobfuscatedClassDumpRecord(
+                            record, proguardMapping, hprofStringCache, classNames, id
+                          )
+                          id += recordsToWrite.size
+                          recordsToWrite.forEach {
+                            writer.write(it)
+                          }
+                        }
+                        else -> writer.write(record)
+                      }
+                    }
+                    else -> writer.write(record)
+                  }
+                }
+                else -> writer.write(record)
+              }
+            })
+            }
+      }
+
+    return outputHprofFile
+  }
+
+  private fun createDeobfuscatedStringRecord(
+    record: StringRecord,
+    proguardMapping: ProguardMapping,
+    hprofStringCache: MutableMap<Long, String>
+  ): StringRecord {
+    val obfuscatedName = hprofStringCache[record.id]!!
+    return StringRecord(
+      record.id, proguardMapping.deobfuscateClassName(obfuscatedName)
+    )
+  }
+
+  /**
+   * Deobfuscated ClassDumpRecord's field names. Different classes can have fields with the same
+   * names. We need to generate new StringRecords in such cases.
+   *
+   * @return a list of HprofRecords to write.
+   */
+  private fun createDeobfuscatedClassDumpRecord(
+    record: ClassDumpRecord,
+    proguardMapping: ProguardMapping,
+    hprofStringCache: MutableMap<Long, String>,
+    classNames: MutableMap<Long, Long>,
+    maxId: Long
+  ): List<HprofRecord> {
+    val recordsToWrite = mutableListOf<HprofRecord>()
+
+    var id = maxId
+
+    val newFields = record.fields.map { field ->
+      val className = hprofStringCache[classNames[record.id]]!!
+      val fieldName = hprofStringCache[field.nameStringId]!!
+      val deobfuscatedName =
+        proguardMapping.deobfuscateFieldName(className, fieldName)
+
+      val newStringRecord = StringRecord(id++, deobfuscatedName)
+      recordsToWrite.add(newStringRecord)
```

Comment:
Should we really create a new record if the field name could not be deobfuscated and we have `fieldName == deobfuscatedName` ?

Reply:
Good point. I think we shouldn't. It unnecessarily makes the hprof size bigger.

Reply:
Changed it